### PR TITLE
Fix mem scramble test

### LINKED ISSUE #14841: [test-triage] chip_sw_otbn_mem_scramble
### Hierarchy of regression failure

Chip Level

### Failure Description

  UVM_ERROR @ 4600.294014 us: (sw_logger_if.sv:521) [otbn_mem_scramble_test_prog_sim_dv(sw/device/tests/otbn_mem_scramble_test.c:211)] CHECK-fail: Expecting at least 49 DMEM integrity errors, got 0

This may be fixed now: I just ran the test at HEAD for 6 iterations with no failures.

### Steps to Reproduce

- Commit hash: https://github.com/lowrisc/opentitan/tree/b73727cd502c1a66545c7f84d7861a1474342963
- dvsim invocation command to reproduce the failure, inclusive of build and run seeds:
  ./util/dvsim/dvsim.py hw/top_earlgrey/dv/chip_sim_cfg.hjson -i chip_sw_otbn_mem_scramble --fixed-seed 1596012791 --build-seed 2341666946 --waves -v h

### Tests with similar or related failures

_No response_

### PR DESCRIPTION
**Pull request by engdoreis**
This PR adjusts the false positive tolerance for ECC error on otbn and sram mem scramble test.
This is the rationale discussed with @vrozic

Each word has 7 bits of ECC, so the probability that a word would generate an ECC error is (127/128). Then the probability that all `kWordsAmount` will generate ECC error is `pow(127/128, kWordsAmount)` after re-scrambling.

Based on that this PR increases the tolerance to false positive in order to fail less in the nightly regression. 

Fixes #14841

### CODE REVIEW

**Comment by tjaychen** [comment]
ah is this similar to the other test where we can't rely on there being an ECC error when the scramble key is just changed or something?

**Comment by engdoreis** [comment]
> ah is this similar to the other test where we can't rely on there being an ECC error when the scramble key is just changed or something?

Yes, the rationale apply for both, so that's why this PR change both tests.

**Comment by tjaychen** [comment]
> > ah is this similar to the other test where we can't rely on there being an ECC error when the scramble key is just changed or something?
> 
> Yes, the rationale apply for both, so that's why this PR changes both tests.

sounds good, thanks for fixing.

File: sw/device/tests/sim_dv/sram_ctrl_main_scrambled_access_test.c [hunk 1]
```diff
@@ -24,11 +24,48 @@
 #define BACKDOOR_TEST_BYTES BACKDOOR_TEST_WORDS * sizeof(uint32_t)
 
 /**
- * Define the maximum false positives the test will tolerate to 1% rounded up.
+ * Note that each word has 7 bits of ECC, so the probability that a word would
```
Comment (matutem) [inline sw/device/tests/sim_dv/sram_ctrl_main_scrambled_access_test.c:27]: Same as above.

**Review by matutem — CHANGES_REQUESTED** [review]
You may want to check with someone with a deeper understanding of SECDED to make sure the comment is accurate. It may just provide a lower bound?

**Review by moidx — APPROVED** [review]

File: sw/device/tests/otbn_mem_scramble_test.c [hunk 2]
```diff
@@ -22,19 +22,59 @@ typedef dif_result_t (*otbn_write_t)(const dif_otbn_t *otbn,
...
+  /**
+   * Number of distinct addresses to check in each IMEM and DMEM.
+   */
+  kNumAddrs = 50,
+
+  /**
+   * Minimum number of expected integrity errors in IMEM and DMEM after
+   * re-scrambling.
+   *
+   * Note that each word has 7 bits of ECC, so the probability that a word would
```
[the remark this answers is not in the document]
Comment (vrozic) [reply sw/device/tests/otbn_mem_scramble_test.c:35]: For what it's worth, I just looked into implementation of `prim_secded_39_32_dec.sv`.
There are certainly some multi-bit errors that will trigger a single-bit error or a double-bit error.
However, everything that is not a valid code word will trigger one of the two errors.

Since we are working with 39-bit words and there are 2^32 valid code words, the probability that a randomly selected 39-bit word triggers an error is indeed 1 in 128.
If the test doesn't care which type of error is triggered, then the remaining calculations look correct to me.

[the remark this answers is not in the document]
Comment (tjaychen) [reply sw/device/tests/otbn_mem_scramble_test.c:35]: somewhat related, there was an old PR that described with valid code words and the "halo" of detection around it. See [here](https://github.com/lowRISC/opentitan/pull/3899#issuecomment-716799810).

I think what @vrozic is saying makes sense to me as well.

Comment (vrozic) [inline sw/device/tests/otbn_mem_scramble_test.c:35]: To be more precise, I would rephrase this comment into something like:
`Note that there are 2^32 valid code words and that each non-valid code word triggers an error. Therefore, the probability that a random 39-bit word triggers an error is: (2^39 - 2^32)/ 2^39 = 127/128.`

The rest looks OK to me.

**Review by vrozic — APPROVED** [review]
Thanks for this PR.
This looks good to me.

File: sw/device/tests/otbn_mem_scramble_test.c [hunk 3]
```diff
@@ -22,19 +22,59 @@ typedef dif_result_t (*otbn_write_t)(const dif_otbn_t *otbn,
...
+   * ```
+   * Pr(0): 0.6756,   sum{Pr(0-0)}: 0.675597
+   * Pr(1): 0.266,    sum{Pr(0-1)}: 0.94158
+   * Pr(2): 0.0513,   sum{Pr(0-2)}: 0.992891
+   * Pr(3): 0.0065,   sum{Pr(0-3)}: 0.999356
+   * Pr(4): 0.0006,   sum{Pr(0-4)}: 0.999954
+   * ```
+   * So by choosen `48 = (kNumAddrs - 2)` as the threshold will have the
+   * probability of `1 - 0.992891 = 0.71%` that this test will fail randomly due
+   * to ECC errors not generated. Which seems a reasonable number.
```
[the remark this answers is not in the document]
Comment (luismarques) [reply sw/device/tests/otbn_mem_scramble_test.c:71]: If we have several tests like this, a 0.71% probability of failure for each test can still lead to a flaky CI. Ideally, we'd make that number lower, but I'm not sure how feasible that is. In some cases there are also ways to work around the problem like using a fixed seed (or set of seeds) for the tests.

[the remark this answers is not in the document]
Comment (engdoreis) [reply sw/device/tests/otbn_mem_scramble_test.c:71]: I don't see a problem in making this number bigger and make the test more robust, I just proposed theses numbers because I was thinking about the nightly regression that runs once a day, which would probably fails once a year moreless. But if we start to run these tests on CI, 0.71% would be anoing.

**Review by matutem — APPROVED** [review]
Thanks for the clarification.
LGTM
